app/api/answers/generate_task_token.go

```diff
@@ -0,0 +1,202 @@
+// responses:
+//   "200":
+//     description: "OK. Success response with the task token"
+//     schema:
+//       type: object
+//       required: [success, message, data]
+//       properties:
+//         success:
+//           description: "true"
+//           type: boolean
+//           enum: [true]
+//         message:
+//           description: updated
+//           type: string
+//           enum: [updated]
+//         data:
+//           type: object
+//           required: [task_token]
+//           properties:
+//             task_token:
+//               type: string
+//   "400":
+//     "$ref": "#/responses/badRequestResponse"
+//   "401":
+//     "$ref": "#/responses/unauthorizedResponse"
+//   "403":
+//     "$ref": "#/responses/forbiddenResponse"
+//   "500":
+//     "$ref": "#/responses/internalErrorResponse"
+func (srv *Service) generateTaskToken(w http.ResponseWriter, r *http.Request) service.APIError {
+	answerID, err := service.ResolveURLQueryPathInt64Field(r, "answer_id")
+	if err != nil {
+		return service.ErrInvalidRequest(err)
+	}
+
+	user := srv.GetUser(r)
+
+	var answerInfos struct {
+		AccessSolutions   bool
+		TextID            *string
+		URL               string
+		SupportedLangProg *string
+		AuthorID          int64
+		AuthorLogin       *string
+		ParticipantID     int64
+		AttemptID         int64
+		ItemID            int64
+		HintsRequested    *string
+		HintsCachedCount  int32 `gorm:"column:hints_cached"`
+		Validated         bool
+	}
+
+	store := srv.GetStore(r)
+	usersGroupsQuery := store.ActiveGroupGroups().WhereUserIsMember(user).Select("parent_group_id")
+
+	// a participant should have at least 'content' access to the answers.item_id
+	participantItemPerms := store.Permissions().MatchingUserAncestors(user).
+		WherePermissionIsAtLeast("view", "content").
+		Where("permissions.item_id = answers.item_id").
+		Select("1").
+		Limit(1)
+
+	// an observer should have 'can_watch'>='answer' permission on the answers.item_id
+	observerItemPerms := store.Permissions().MatchingUserAncestors(user).
+		WherePermissionIsAtLeast("watch", "answer").
+		Where("permissions.item_id = answers.item_id").
+		Select("1").
+		Limit(1)
+
+	// an observer should be able to watch the participant
+	observerParticipantPerms := store.ActiveGroupAncestors().ManagedByUser(user).
+		Joins("JOIN `groups` ON groups.id = groups_ancestors_active.child_group_id").
+		Where("groups_ancestors_active.child_group_id = answers.participant_id").
+		Where("can_watch_members").
+		Select("1").
+		Limit(1)
+
+	err = store.Answers().WithItems().WithResults().
+		Joins("JOIN groups_ancestors_active ON groups_ancestors_active.child_group_id = ?", user.GroupID).
+		Joins(`
+				JOIN permissions_generated
+					ON permissions_generated.item_id = items.id AND
+						 permissions_generated.group_id = groups_ancestors_active.ancestor_group_id`).
+		Joins("JOIN attempts ON attempts.participant_id = results.participant_id AND attempts.id = results.attempt_id").
```

Comment:
do you use the `attempts` table for something?

Reply:
fixed